### PR TITLE
PIM-5640: Apply permissions on quick export

### PR DESCRIPTION
**Pull request by phaseinducer**
**What does this PR do?**
- Quick export processor is now user aware (like the mass edit processor for example)
- Product normalizers (structured & flat) now accept several filters in context

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | yes
| Added Behats                      | n/a
| Changelog updated                 | yes
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) | in progress
| Migration script                  | n/a
| Tech Doc                          |

### CODE REVIEW

File: src/Pim/Component/Connector/Normalizer/Flat/ProductNormalizer.php [hunk 1]
```diff
@@ -246,6 +224,14 @@ protected function normalizeAssociations($associations = [])
      */
     protected function resolveContext(array $context)
     {
-        return array_merge(['scopeCode' => null, 'localeCodes' => []], $context);
+        return array_merge(
+            [
+                'scopeCode'     => null,
+                'localeCodes'   => [],
+                'metric_format' => 'multiple_fields',
+                'filter_types'  => ['pim.transform.product_value.flat']
```
Comment (solivier) [inline src/Pim/Component/Connector/Normalizer/Flat/ProductNormalizer.php]: maybe pass it via constructor ?

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductNormalizer.php [hunk 2]
```diff
@@ -68,6 +68,7 @@ public function normalize($product, $format = null, array $context = [])
         $defaultContext = [
             'only_associations'    => false,
             'exclude_associations' => false,
+            'filter_types'         => ['pim.transform.product_value.structured']
```
Comment (solivier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductNormalizer.php]: maybe pass it via constructor ?

File: spec/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessorSpec.php [hunk 3]
```diff
@@ -271,6 +336,11 @@ function it_returns_flat_data_with_french_attribute(
         ProductPriceInterface $price,
         ProductValueInterface $priceValue
     ) {
+        $stepExecution->getJobExecution()->willReturn($jobExecution);
+        $jobExecution->getUser()->willReturn('michel');
+        $userProvider->loadUserByUsername('michel')->willReturn($user);
+        $user->getRoles()->willReturn(['ROLE_MICHEL']);
```
Comment (solivier) [inline spec/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessorSpec.php]: :+1:

Comment (juliensnz) [inline spec/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessorSpec.php]: 👍

**Comment by solivier** [comment]
gtm

**Comment by SamirBoulil** [comment]
gtm

File: src/Pim/Component/Connector/Normalizer/Flat/ProductNormalizer.php [hunk 4]
```diff
@@ -106,33 +97,17 @@ protected function normalizeProperties(ProductInterface $product)
...
-                    || isset($this->fields[$fieldValue])) {
-                    $normalizedValue = $this->serializer->normalize($value, $format, $context);
-                    $this->results = array_replace($this->results, $normalizedValue);
-                }
-            }
+        $values = $this->getFilteredValues($product, $context);
+
+        $normalizedValues = [];
+        foreach ($values as $value) {
+            $normalizedValues = array_replace(
```
Comment (a2xchip) [inline src/Pim/Component/Connector/Normalizer/Flat/ProductNormalizer.php]: What is the purpose of this loop?

Comment (phaseinducer) [inline src/Pim/Component/Connector/Normalizer/Flat/ProductNormalizer.php]: The ProductValue normalizer returns something like `['attribute_code' => 'normalized value']`, so it's just a convenient way to add the new normalized value in `$normalizedValues` (which is also indexed by attribute codes).

**Pull request merged**